File: src/examples/ex_pos_ctrl.cpp
/* ex_pos_ctrl.cpp  (MJCF)
 * Joint position control: drive the Kinova GEN3 from the home pose to a target
 * pose using linearly interpolated position setpoints, then hold.
 *
 * The setpoint trajectory is a straight-line interpolation in joint space over
 * kMotionDuration seconds.  After the motion completes the final position is
 * held for 1 s.
 *
 * Requires MuJoCo Menagerie in cache.
 *
 * Usage:
 *   ex_pos_ctrl [--headless]
 *
 * Runs the full motion once and exits; --headless skips the viewer and prints the final
 * max joint error. */

#include "mj_kdl_wrapper/mj_kdl_wrapper.hpp"
#include "common.hpp"
#include "example_paths.hpp"

#include <algorithm>
#include <cmath>
#include <iomanip>
#include <iostream>
#include <string>

using mj_kdl_examples::kHomePose;
static constexpr double kTargetPose[7]  = { 0.3, 0.5, 2.9, -2.0, 0.3, 1.2, 1.3 };
static constexpr double kMotionDuration = 2.0; // seconds for the interpolated move

int main(int argc, char *argv[])
{
    const bool headless = mj_kdl_examples::parse_args(argc, argv).headless;

    const std::string mjcf = mj_kdl_examples::menagerie_model("kinova_gen3/gen3.xml");

    mj_kdl::SceneSpec sc;
    sc.timestep   = 0.002;
    sc.add_floor  = true;
    sc.add_skybox = true;
    mj_kdl::RobotSpec r;
    r.path = mjcf;
    sc.robots.push_back(r);

    mj_kdl::Env env;
    if (!mj_kdl::init_env(&env, &sc)) {
        std::cerr << "init_env() failed\n";
        return 1;
    }

    mj_kdl::Robot robot;
    if (!mj_kdl::init_robot_from_mjcf(&robot, &env, "base_link", "bracelet_link")) {
        std::cerr << "init_robot_from_mjcf() failed\n";
        return 1;
    }

    unsigned n = static_cast<unsigned>(robot.n_joints);

    KDL::JntArray q_home(n);
    for (unsigned i = 0; i < n; ++i) q_home(i) = kHomePose[i];

    if (!mj_kdl::set_control_mode(&robot, mj_kdl::CtrlMode::POSITION)) return 1;

    double t_start = 0.0;

    env.on_reset = [&](mj_kdl::ResetContext *ctx) {
        mj_kdl::set_joint_pos(&robot, q_home);
        t_start = ctx->data->time;
    };

    mj_kdl::reset(&env);

    auto ctrl_step = [&]() {
        mj_kdl::update(&env);

        double alpha = std::clamp((env.data->time - t_start) / kMotionDuration, 0.0, 1.0);
        for (unsigned i = 0; i < n; ++i)
            robot.jnt_pos_cmd[i] = kHomePose[i] + alpha * (kTargetPose[i] - kHomePose[i]);
    };

    // Run until the trajectory is complete plus a short settling period.
    const double end_time = kMotionDuration + 1.0;
    if (headless) {
        while (env.data->time < end_time) {
            ctrl_step();
            mj_kdl::step(&env);
        }

        double max_err = 0.0;
        for (unsigned i = 0; i < n; ++i)
            max_err = std::max(max_err, std::abs(kTargetPose[i] - robot.jnt_pos_msr[i]));
        std::cout << "max joint error at end: " << std::fixed << std::setprecision(4) << max_err
                  << " rad\n";
    } else {
        if (!mj_kdl::open_viewer(&env)) {
            std::cerr << "open_viewer() failed\n";
            return 1;
        }
        while (env.data->time - t_start < end_time) {
            ctrl_step();
            if (!mj_kdl::step(&env)) break;
            mj_kdl::pace_realtime(&env);
        }
    }

    mj_kdl::cleanup(&env);
    return 0;
}

File: python/examples/ex_pos_ctrl.py
#!/usr/bin/env python3
"""Joint position-control example ported from src/examples/ex_pos_ctrl.cpp.

Drives the arm from home to a target with linearly interpolated position
setpoints. An Env on_reset hook re-homes the arm and restarts the motion clock
so the simulate-UI reset button replays the motion.
"""

from __future__ import annotations

import argparse
import os
from pathlib import Path

import mj_kdl_wrapper as mjk

DEFAULT_MODEL = "third_party/menagerie/kinova_gen3/gen3.xml"
MODEL_ENV_VAR = "MJ_KDL_MODEL"

HOME_POSE = [0.0, 0.2618, 3.1416, -2.2689, 0.0, 0.9599, 1.5708]
TARGET_POSE = [0.3, 0.5, 2.9, -2.0, 0.3, 1.2, 1.3]
MOTION_DURATION = 2.0


def clamp(value: float, low: float, high: float) -> float:
    return max(low, min(high, value))


def build_env(model_path: Path) -> tuple[mjk.Env, mjk.Robot]:
    spec = mjk.SceneSpec()
    spec.timestep = 0.002
    spec.add_floor = True
    spec.add_skybox = True
    robot_spec = mjk.RobotSpec()
    robot_spec.path = str(model_path)
    spec.robots = [robot_spec]
    env = mjk.Env.build(spec)
    robot = env.create_robot("base_link", "bracelet_link")
    robot.ctrl_mode = mjk.CtrlMode.POSITION
    return env, robot


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--gui", action="store_true", help="Open the custom Simulate UI.")
    args = parser.parse_args()

    model_path = Path(os.environ.get(MODEL_ENV_VAR, DEFAULT_MODEL))
    if not model_path.exists():
        raise FileNotFoundError(
            f"{model_path} does not exist. Run from a directory where that relative path exists "
            f"or set {MODEL_ENV_VAR}."
        )

    env, robot = build_env(model_path)
    try:
        # t_start is reset on every env reset so the trajectory replays from home.
        state = {"t_start": 0.0}

        def on_reset(ctx):
            robot.set_joint_pos(HOME_POSE, call_forward=False)
            robot.jnt_pos_cmd = HOME_POSE[:]
            state["t_start"] = env.time()

        env.on_reset = on_reset
        env.reset()

        def control_step():
            robot.update()
            alpha = clamp((env.time() - state["t_start"]) / MOTION_DURATION, 0.0, 1.0)
            robot.jnt_pos_cmd = [
                HOME_POSE[i] + alpha * (TARGET_POSE[i] - HOME_POSE[i])
                for i in range(robot.n_joints)
            ]

        if args.gui:
            viewer = mjk.SimulateViewer.open(robot, "ex_pos_ctrl.py")
            prev = env.time()
            try:
                while viewer.is_running():
                    if env.time() < prev - 1e-6:
                        env.reset()
                    prev = env.time()
                    control_step()
                    if not viewer.step():
                        break
            finally:
                viewer.close()
        else:
            end = env.time() + MOTION_DURATION + 1.0
            while env.time() < end:
                control_step()
                robot.step()
            max_err = max(abs(TARGET_POSE[i] - robot.jnt_pos_msr[i]) for i in range(robot.n_joints))
            print(f"max joint error at end: {max_err:.4f} rad")
    finally:
        env.close()

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
